File: Park/CodeTree/가중치가 동일한 그래프에서의 BFS/최소 경로로 탈출 하기.py
import sys
sys.setrecursionlimit(10000)  # 재귀 한도 설정

INT_MAX = sys.maxsize

# 변수 선언 및 입력
n, m = map(int, input().split())

a = [list(map(int, input().split())) for _ in range(n)]

visited = [[False] * m for _ in range(n)]  # 방문 여부 체크

ans = INT_MAX  # 최소 이동 횟수 저장

# 좌표가 격자 내에 있는지 확인
def in_range(x, y):
    return 0 <= x < n and 0 <= y < m

# 이동 가능 여부 확인
def can_go(x, y):
    return in_range(x, y) and a[x][y] and not visited[x][y]

# 최소 이동 횟수를 찾는 함수
def find_min(x, y, cnt):
    global ans
    
    if x == n - 1 and y == m - 1:  # 도착점에 도달한 경우
        ans = min(ans, cnt)
        return
    
    # 4방향 이동 (우, 하, 좌, 상)
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    for dx, dy in directions:
        new_x, new_y = x + dx, y + dy
        
        if can_go(new_x, new_y):
            visited[new_x][new_y] = True
            find_min(new_x, new_y, cnt + 1)
            visited[new_x][new_y] = False  # 백트래킹

# 초기 위치에서 시작
visited[0][0] = True
find_min(0, 0, 0)

# 불가능한 경우 -1 출력
if ans == INT_MAX:
    ans = -1

print(ans)

File: jhyun/CodeTree/BFS/가중치가 동일한 그래프에서의 BFS/최소_경로로_탈출_하기.java
import java.util.*;
import java.io.*;

class Pair {
    int x, y;

    public Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class Main {
    static final int INT_MAX = Integer.MAX_VALUE;
    static final int MAX_N = 100;
    static final int MAX_M = 100;

    static int n, m;
    static int[][] step = new int[MAX_N][MAX_M];
    static int[][] grid = new int[MAX_N][MAX_M];
    static boolean[][] visited = new boolean[MAX_N][MAX_M];
    static Queue<Pair> queue = new LinkedList<>();

    static int ans = INT_MAX;

    static boolean inRange(int x, int y) {
        return 0 <= x && x < n && 0 <= y && y < m;
    }

    static boolean canGo(int x, int y) {
        return inRange(x, y) && grid[x][y] == 1 && !visited[x][y];
    }

    static void push(int x, int y, int s) {
        queue.add(new Pair(x, y));
        visited[x][y] = true;
        step[x][y] = s;
    }

    static void findMin() {
        while (!queue.isEmpty()) {
            Pair curr = queue.poll();
            int x = curr.x;
            int y = curr.y;

            int[] dx = new int[]{0, 1, 0, -1};
            int[] dy = new int[]{1, 0, -1, 0};

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];

                if (canGo(newX, newY)) {
                    push(newX, newY, step[x][y] + 1);
                }
            }
        }

        if (visited[n - 1][m - 1]) {
            ans = step[n - 1][m - 1];
        }
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                grid[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        push(0, 0, 0);
        findMin();

        if (ans == INT_MAX) {
            ans = -1;
        }

        System.out.println(ans);
    }
}